File: content/exercises/r-for-economists/reproducible-research/regression-table/checker.R
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 2) {
  stop("Usage: checker.R <submission_path> <output_path>")
}

submission_path <- args[[1]]
output_path <- args[[2]]

test_results <- list()
feedback <- c()
status <- "failed"

record_test <- function(name, passed, message) {
  test_results[[length(test_results) + 1]] <<- list(
    name = name,
    passed = passed,
    message = message
  )
}

env <- new.env(parent = baseenv())
execution_ok <- TRUE
execution_error <- NULL

tryCatch(
  {
    source(submission_path, local = env)
  },
  error = function(e) {
    execution_ok <<- FALSE
    execution_error <<- conditionMessage(e)
  }
)

if (!execution_ok) {
  record_test("code runs", FALSE, "Your code did not run successfully.")
  feedback <- c(feedback, paste("Execution error:", execution_error))
} else {
  record_test("code runs", TRUE, "Code executed without errors.")
  for (model_name in c("m1", "m2", "m3")) {
    exists_ok <- exists(model_name, envir = env, inherits = FALSE)
    record_test(
      paste0(model_name, " exists"),
      exists_ok,
      paste0("Define model `", model_name, "`.")
    )
    if (exists_ok) {
      model <- get(model_name, envir = env, inherits = FALSE)
      is_lm <- inherits(model, "lm")
      record_test(
        paste0(model_name, " class"),
        is_lm,
        paste0("`", model_name, "` should be an lm model.")
      )
    }
  }
}

passed_all <- length(test_results) > 0 && all(vapply(test_results, function(t) isTRUE(t$passed), logical(1)))
if (passed_all) {
  status <- "passed"
  feedback <- c(feedback, "Correct solution submitted.")
} else if (length(feedback) == 0) {
  feedback <- c(feedback, "Fit m1, m2, m3 and call modelsummary(list(m1, m2, m3), stars = TRUE).")
}

output_lines <- c(paste0("STATUS:", status))
for (msg in feedback) {
  output_lines <- c(output_lines, paste0("FEEDBACK:", msg))
}
for (t in test_results) {
  pass_label <- if (isTRUE(t$passed)) "pass" else "fail"
  output_lines <- c(output_lines, paste0("TEST:", t$name, "|", pass_label, "|", t$message))
}

writeLines(output_lines, con = output_path)

File: content/exercises/economics-data-science/causal-inference/diff-in-diff/checker.py
import io
import runpy
import sys


def main() -> None:
    submission_path = sys.argv[1]
    output_path = sys.argv[2]
    tests: list[tuple[str, bool, str]] = []
    feedback: list[str] = []
    status = "failed"

    try:
        namespace = runpy.run_path(submission_path, run_name="__main__")
        tests.append(("code runs", True, "Code executed without errors."))
    except Exception as error:  # noqa: BLE001
        namespace = {}
        tests.append(("code runs", False, "Code did not run successfully."))
        feedback.append(f"Execution error: {error}")

    df = namespace.get("df")
    if df is None:
        tests.append(("df exists", False, "Define DataFrame variable df."))
    else:
        has_col = "treated_post" in getattr(df, "columns", [])
        tests.append(("treated_post column", has_col, "Create df['treated_post'] = treated * post."))

    model = namespace.get("model")
    if model is None:
        tests.append(("model exists", False, "Fit OLS model and store it in model."))
    else:
        params = getattr(model, "params", {})
        has_param = "treated_post" in getattr(params, "index", []) or "treated_post" in params
        tests.append(("did parameter", has_param, "Model should include treated_post coefficient."))

    passed_all = tests and all(item[1] for item in tests)
    if passed_all:
        status = "passed"
        feedback.append("Correct solution submitted.")
    elif not feedback:
        feedback.append("Create treated_post and estimate the DiD OLS model.")

    output_stream = io.StringIO()
    output_stream.write(f"STATUS:{status}\n")
    for item in feedback:
        output_stream.write(f"FEEDBACK:{item}\n")
    for name, passed, message in tests:
        output_stream.write(f"TEST:{name}|{'pass' if passed else 'fail'}|{message}\n")
    with open(output_path, "w", encoding="utf-8") as handle:
        handle.write(output_stream.getvalue())


if __name__ == "__main__":
    main()
